README.md
# Numerical_Methods_II

Exc3/HISTORY
amber                                                                           
         0         2        18
timestep       100        18         0         2    0.005000
   25.20       0.000       0.000    
   0.000       25.20       0.000    
   0.000       0.000       25.20    
OW               1   15.999400   -0.847600
 -6.7467E+00  1.0891E+01  6.4941E+00
HW               2    1.007900    0.423800
 -6.2004E+00  1.1662E+01  6.3384E+00
HW               3    1.007900    0.423800
 -6.9578E+00  1.0931E+01  7.4270E+00
OW               4   15.999400   -0.847600
 -3.5090E+00  1.0291E+01  8.8284E+00
HW               5    1.007900    0.423800
 -4.0539E+00  1.1078E+01  8.8520E+00
HW               6    1.007900    0.423800
 -3.6673E+00  9.8639E+00  9.6704E+00
OW               7   15.999400   -0.847600
 -8.1353E+00  6.7479E+00  8.6033E+00
HW               8    1.007900    0.423800
 -7.5158E+00  7.4467E+00  8.8134E+00
HW               9    1.007900    0.423800
 -7.6144E+00  6.0998E+00  8.1291E+00
OW              10   15.999400   -0.847600
 -5.3982E+00 -1.0612E+01  6.2198E+00
HW              11    1.007900    0.423800
 -5.3418E+00 -1.0387E+01  5.2910E+00
HW              12    1.007900    0.423800
 -6.3246E+00 -1.0506E+01  6.4364E+00
OW              13   15.999400   -0.847600
 -4.4865E+00  7.5880E+00  1.2439E+01
HW              14    1.007900    0.423800
 -4.6666E+00  7.5093E+00 -1.1822E+01
HW              15    1.007900    0.423800
 -3.5366E+00  7.4927E+00  1.2368E+01
OW              16   15.999400   -0.847600
 -5.3320E+00 -1.2038E+01  1.0083E+01
HW              17    1.007900    0.423800
 -5.6750E+00 -1.2565E+01  1.0804E+01
HW              18    1.007900    0.423800
 -4.3867E+00 -1.2187E+01  1.0105E+01
timestep       200        18         0         2    0.005000
   25.20       0.000       0.000    
   0.000       25.20       0.000    
   0.000       0.000       25.20    
OW               1   15.999400   -0.847600
 -3.6487E+00 -1.2471E+01  5.1296E+00
HW               2    1.007900    0.423800
 -3.8723E+00  1.1843E+01  5.4222E+00
HW               3    1.007900    0.423800
 -4.4708E+00 -1.1984E+01  5.1868E+00
OW               4   15.999400   -0.847600
 -4.8343E+00  6.7317E+00  6.9344E+00
HW               5    1.007900    0.423800
 -4.0340E+00  7.0167E+00  7.3756E+00
HW               6    1.007900    0.423800
 -5.5320E+00  7.2350E+00  7.3542E+00
OW               7   15.999400   -0.847600
 -8.9648E+00  5.6523E+00  1.1021E+01
HW               8    1.007900    0.423800
 -8.5610E+00  4.8845E+00  1.0616E+01
HW               9    1.007900    0.423800
 -9.3056E+00  5.3320E+00  1.1856E+01
OW              10   15.999400   -0.847600
 -3.4924E+00 -6.7683E+00  1.5490E+00
HW              11    1.007900    0.423800
 -4.1783E+00 -7.3306E+00  1.1889E+00
HW              12    1.007900    0.423800
 -2.6740E+00 -7.1866E+00  1.2815E+00
OW              13   15.999400   -0.847600
 -3.5181E+00  6.2567E+00 -5.9004E+00
HW              14    1.007900    0.423800
 -2.9606E+00  5.8261E+00 -5.2522E+00
HW              15    1.007900    0.423800
 -3.8485E+00  5.5423E+00 -6.4452E+00
OW              16   15.999400   -0.847600
 -9.0047E+00 -1.1575E+01  8.8268E+00
HW              17    1.007900    0.423800
 -9.1826E+00 -1.0674E+01  9.0952E+00
HW              18    1.007900    0.423800
 -9.6175E+00 -1.1737E+01  8.1096E+00
timestep       300        18         0         2    0.005000
   25.20       0.000       0.000    
   0.000       25.20       0.000    
   0.000       0.000       25.20    
OW               1   15.999400   -0.847600
 -2.4328E+00 -1.0701E+01  4.1909E+00
HW               2    1.007900    0.423800
 -1.9806E+00 -1.1543E+01  4.1360E+00
HW               3    1.007900    0.423800
 -2.9235E+00 -1.0745E+01  5.0116E+00
OW               4   15.999400   -0.847600
 -6.0000E+00  3.6975E+00  5.6492E+00
HW               5    1.007900    0.423800
 -5.6017E+00  2.9109E+00  5.2763E+00
HW               6    1.007900    0.423800
 -6.1470E+00  3.4789E+00  6.5695E+00
OW               7   15.999400   -0.847600
 -9.2878E+00  4.4619E+00 -1.2191E+01
HW               8    1.007900    0.423800
 -9.7002E+00  3.7679E+00 -1.1676E+01
HW               9    1.007900    0.423800
 -8.8343E+00  5.0028E+00 -1.1544E+01
OW              10   15.999400   -0.847600
 -1.6987E+00 -3.4865E+00 -2.4544E+00
HW              11    1.007900    0.423800
 -1.4579E+00 -2.6683E+00 -2.8891E+00
HW              12    1.007900    0.423800
 -1.8251E+00 -3.2428E+00 -1.5373E+00
OW              13   15.999400   -0.847600
 -2.6243E+00  4.9498E+00  1.1768E-01
HW              14    1.007900    0.423800
 -2.2785E+00  4.9194E+00  1.0098E+00
HW              15    1.007900    0.423800
 -3.5519E+00  5.1592E+00  2.2681E-01
OW              16   15.999400   -0.847600
 -1.1459E+01 -1.1106E+01  6.8748E+00
HW              17    1.007900    0.423800
 -1.1540E+01 -1.1591E+01  6.0539E+00
HW              18    1.007900    0.423800
 -1.1768E+01 -1.1715E+01  7.5456E+00
timestep       400        18         0         2    0.005000
   25.20       0.000       0.000    
   0.000       25.20       0.000    
   0.000       0.000       25.20    
OW               1   15.999400   -0.847600
 -1.2679E+00 -8.8900E+00  3.3898E+00
HW               2    1.007900    0.423800
 -1.1628E+00 -8.7620E+00  2.4470E+00
HW               3    1.007900    0.423800
 -1.4443E+00 -9.8258E+00  3.4869E+00
OW               4   15.999400   -0.847600
 -7.0378E+00  7.4425E-01  4.5019E+00
HW               5    1.007900    0.423800
 -7.0569E+00 -1.0039E-01  4.9519E+00
HW               6    1.007900    0.423800
 -6.8380E+00  5.2663E-01  3.5914E+00
OW               7   15.999400   -0.847600
 -9.6184E+00  3.2706E+00 -1.0081E+01
HW               8    1.007900    0.423800
 -8.7798E+00  3.4456E+00 -1.0508E+01
HW               9    1.007900    0.423800
 -9.6847E+00  2.3158E+00 -1.0064E+01
OW              10   15.999400   -0.847600
  1.5357E-01 -9.3770E-03 -6.4208E+00
HW              11    1.007900    0.423800
  1.0877E+00  1.6700E-01 -6.5333E+00

exercise3.py
'''
Numerical Methods II
Exercise 3
Lecture 5
Integral with Importance Sampling using Metropolis Algorithm
'''

__author__ = "JL Euste"


import numpy as np
import matplotlib.pyplot as plt
from scipy.special import gamma,erf

np.random.seed(11)

sigmas = [0.1,1,10]

dim = 4 #dimensionality of the integral
solid_angle = 2*np.pi**(dim/2.)/gamma(dim/2.) #will be used from Cartesian to 4d spherical

def metropolis_integral(sigma,h,cyc=100,N_wup=5_000,N_meas=10_000,infty=True):
    def g(x):
        norm2 = np.sum(x**2)
        return (np.sin(np.pi*norm2/2))**2

    def p(x):
        norm2 = np.sum(x**2)
        return np.exp(-norm2/sigma)
    

    def to_update(x0,x1): #defines if we update or not
        a = min(1.,p(x1)/p(x0))
        
        if (a>=1.):  return x1 #accept
        else:
            r = np.random.random()
            if (r>a):   return x0 #reject
            else:   return x1 #accept
    
    if infty: #from -infinity to +infinity
        def markov_chain(x_in,cc): #Markov chain cycle
            for _ in range(cc):
                x_out = x_in + h*s_val*np.random.randn(dim)
                x_in = to_update(x_in,x_out)
            return x_in
        
        norm_fac = np.sqrt(np.pi*sigma)**dim #from normalization of p(x)
        
        #actual values from wolframalpha
        actual_vals = {0.1: 0.000633379*solid_angle,
                       1: 0.2687679475993*solid_angle,
                       10: 25.0252534*solid_angle} 
    else: #set infty to False if from 0 to 4
        def markov_chain(x_in,cc): #Markov chain cycle
            for _ in range(cc):
                x_out = x_in + h*np.random.randn(dim)/1000 #changed so <4
                x_in = to_update(x_in,x_out)
                
                #to ensure value within [0,4]
                if np.any(np.abs(x_in)>4): print('value out of bounds')
            return x_in
        
        norm_fac = (0.5*np.sqrt(np.pi*sigma)*erf(4/np.sqrt(sigma)))**dim #from normalization of p(x)
        
        #actual values from wolframalpha
        #divide by 16 because "quadrant" in 4D space
        actual_vals = {0.1: 0.000633379*solid_angle/16, 
                        1: 0.26876795*solid_angle/16,
                        10: 24.71809246*solid_angle/16}


    s_val = np.sqrt(sigma/2)    
    
    iters = np.arange(N_meas)
    int_res,stdevs = np.empty(N_meas),np.empty(N_meas)
    aa = []
    bb=0
    
    x0 = s_val*np.random.randn(dim) #initial guess
    x0 = markov_chain(x0,N_wup)
    
    for i in iters:
        x0 = markov_chain(x0,cyc)
        
        meas = g(x0)
        aa.append(meas)       	    
        	
        bb += meas #add this to the 'total' function
        int_res[i] = bb/(i+1) #new average value
        
        std = np.std(aa/np.sqrt(i+1)) #stdev
        stdevs[i] = std
        
    
   
    percent_dev = 100*abs(actual_vals[sigma]-norm_fac*bb/(i+1))/actual_vals[sigma]
    
    print(f'sigma={sigma}')
    print('Percent deviation from actual:',percent_dev) #change depending on sigma

    return iters,norm_fac*int_res,stdevs,actual_vals[sigma],percent_dev



# =============================================================================
# Full integral from -infinity to +infinity
# =============================================================================
fig,ax = plt.subplots(3,2,figsize=(12,16))
for i in range(len(sigmas)):
    sigma = sigmas[i]
    iters,int_res,stdevs,actual,percent_dev = metropolis_integral(h=1.5,
                                                                  sigma=sigma)
    ax[i,0].set_title('$\sigma$'+ f'={sigma}, %dev={percent_dev:.2f}')
    ax[i,0].errorbar(iters, int_res,yerr=stdevs,label='calculated')
    ax[i,0].plot(iters,np.full_like(iters,actual),'r--',label='actual')
    ax[i,0].legend()
    ax[i,0].set_ylabel('MC value',size='x-large')
    if i==(len(sigmas)-1): ax[i,0].set_xlabel('step',size='x-large')
    
    ax[i,1].set_title('Error')
    ax[i,1].loglog(iters,stdevs)
    ax[i,1].set_ylabel('$\sigma$',size='x-large')
    if i==(len(sigmas)-1): ax[i,1].set_xlabel('step',size='x-large')

fig.suptitle('Full integral from -inf to +inf')
fig.tight_layout()
plt.savefig('full_int')



# =============================================================================
# Proper integral from 0 to 4
# =============================================================================
fig,ax = plt.subplots(3,2,figsize=(12,16))
for i in range(len(sigmas)):
    h=1.5
    sigma = sigmas[i]
    if (sigma==1): h=0.06 #smaller so values within range [0,4]
    
    iters,int_res,stdevs,actual,percent_dev = metropolis_integral(h=h,
                                                                  sigma=sigma,
                                                                  infty=False)
    ax[i,0].set_title('$\sigma$'+ f'={sigma}, %dev={percent_dev:.2f}')
    ax[i,0].errorbar(iters, int_res,yerr=stdevs,label='calculated')
    ax[i,0].plot(iters,np.full_like(iters,actual),'r--',label='actual')
    ax[i,0].legend()
    ax[i,0].set_ylabel('MC value',size='x-large')
    if i==(len(sigmas)-1): ax[i,0].set_xlabel('step',size='x-large')
        
    ax[i,1].set_title('Error')
    ax[i,1].loglog(iters,stdevs)
    ax[i,1].set_ylabel('$\sigma$',size='x-large')
    if i==(len(sigmas)-1): ax[i,1].set_xlabel('step',size='x-large')

fig.suptitle('Proper integral from 0 to 4')
fig.tight_layout()
plt.savefig('proper_int_0_4')


Plot-Fit_example.py
import math as mt
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit


#parameters we are interested in
P_in = 0.85
L_in = 0.81

# load input
N_l, Ar, Vr = np.loadtxt("Output_%sP_%sL.txt" % (P_in,L_in))

## Fitting procedure is down in two steps:
## Step 1: identify the target function

def func(x, a, b):
    return a * x**b

## Step 2: perform the fit

popt, pcov = curve_fit(func, N_l, Vr)

print("Best fit with power: %s \n and prefactor %s " % (popt[1],  popt[0]))

#print("Best fit with power: %s (%s) \n and prefactor %s (%s)" % (popt[1], pcov[1], popt[0], pcov[0]))



## Plotting routines

plt.loglog(N_l, Vr, 'o', label='Error')
plt.loglog(N_l, func(N_l, *popt), '-', label='Fit (a+x^b), b=%.4f' % popt[1])

plt.legend(loc=1,fontsize=15)
#plt.ticklabel_format(axis='y', style='sci', scilimits=(0,0))

plt.xlabel("Number of points",fontsize=15)
plt.ylabel("Error",fontsize=15)
#plt.semilogx(L_list,At_tau)
plt.savefig('Plot_Fit_%sP_%sL.png' % (P_in,L_in))

hitormiss_Variances.py
'''
Evaluate the Area of a circle of radius 1, pi * r**2.
Using numpy arrays in parallel

Once data are generated, computes variances

Challenge: make the code more efficient / why is it very slow?
'''

import numpy as np
import math as mt
import pylab as pl  # useful to plot stuff!

# number of points we will draw
NN = 1000
# minimum set we consider for our analysis
n_min = 480
# NN rows of data, with two values each, x and y
data = np.random.rand(NN,2)

xs = data[:,0]
ys = data[:,1]
# A boolean array where x^2 + y^2 < 1:
hits = xs**2 + ys**2 < 1.0

Av = []
dA_sq = []
Var = []
Sigma = []

# Computation of the average
A = float(4 * np.count_nonzero(hits)) / NN

#computation of the differences 
for j in range(n_min, NN):
    # Computation of the average at fixed NN	
    AV_n = float(4 * np.count_nonzero(hits[0:j])) / j
    Av.append(AV_n)
    dA = 0
    for k in range(0, j):
    #computation of the 
        dA += (float(4 * np.count_nonzero(hits[k]) - AV_n))**2
    Var.append(dA)
    Sigma.append(mt.sqrt(dA / (j * (j+1))))



Err = Sigma[NN-(n_min+1)]

# area in one quarter
#A = float(4 * np.count_nonzero(hits)) / NN

Exp = (A - mt.pi)

print("Number of points used: %s" % NN)
print("Value obtained: %s" % A)
print("Error versus expected value: %s" % Exp)
print("Estimated error: %s" % Err)

x_vec = np.arange(NN)

pl.title('Standard deviations')

# generate a reference line with 1/\sqrt{x}
xspace = np.linspace(NN/100., NN, NN)
#pl.loglog(1./np.sqrt(xspace),'bd', label='Reference line')

pl.semilogy(Sigma,'r-', label='Stand. Dev.')
pl.semilogy(1./np.sqrt(xspace),'bd', label='Square root scaling')
pl.semilogy(1./(xspace),'g.', label='Linear scaling')

pl.xlabel('step')
pl.ylabel('$\sigma$')
pl.xlim(10, NN)
#pl.ylim(min(sigma_vec), max(sigma_vec))
pl.legend()

pl.savefig('EstimateCirle.png')

pl.show()

Hit_or_miss_convergence.py
'''
Evaluate the Area of a circle of radius 1, pi * r**2.
Using numpy arrays in parallel

Plot the running average to have an idea about convergence
'''

import numpy as np
import math as mt
import matplotlib.pyplot as plt

# number of points we will draw
NN = 1000
# minimal number of points to start our plot
N_Start = int(NN*0.1)
# NN rows of data, with two values each, x and y
data = np.random.rand(NN,2)
# remember: rand has support [0, 1[
# helpful slices
xs = data[:,0]
ys = data[:,1]

# Array of data
Area = []

# A boolean array where x^2 + y^2 < 1:
# condition: is our point inside the circle or outside?
hits = xs**2 + ys**2 < 1.0

# Counter
CC = 0

# computing running average 
for j in range(NN):
    CC += float(4 * np.count_nonzero(hits[j])) 
    Area.append(CC / (j+1))

# this is just a check, since we know that the 
# final result is pi.
Exp = (float(CC/NN) - mt.pi)

print(CC)



print("Number of points used (total pinballs): %s" % NN)
print("Value obtained: %s" % float(CC/NN))
print("Error versus expected value: %s" % Exp)

#plt.plot(Area[N_Start:NN])

#plt.show()

exit()

hitormiss_adaptedSampling.py
'''
Evaluate the integral

\int_0^1 1/x**pp

utilizing adaptive sampling

'''
import numpy as np
import math as mt
import statistics as st
import pylab as pl  # useful to plot stuff!

# number of points we will consider

N_l = [1000, 2000, 4000, 8000, 16000, 32000]

P_in = 0.85
L_in = 0.81

# true value for safety checks
TT = 1. / (1.-P_in)

#print("True value: %s" % TT)

# vectors of results
Ar = []
Vr = []

# defines a function that, given a value of a variable
# and two powers, pp, ll, returns:
# the value of the function

def IntFun(XX, PP, LL):
	return 1./XX**(PP-LL)

# defines a function that, taken a number of points, the value of p, 
# and the value of \ell, returns:
# mean, variance
def Ip_est(NN, pp, ell):
	# p2 normalization
	NP2 = 1. / (1.-ell)
	# here we are using direct sampling of the distribution
	# p = 1/ x**p2
	# so we know that:
	# x  = u**(1/(1-p2))
	datax = (np.random.rand(NN))**(1. / (1.-ell))
	# generation of the result for the estimate
	YY = IntFun(datax,pp, ell)
	# mean
	A=NP2*np.sum(YY)/NN
	# error
	VV = st.variance(YY) / mt.sqrt(NN)
	#print("Number of points used: %s" % NN)
	#print("Value obtained: %s" % A)
	#print("Error: %s" % VV)
	return(A, VV)


for j in N_l:
	aa, vv = Ip_est(j, P_in, L_in)
	Ar.append(aa),
	Vr.append(vv)

np.savetxt("Output_%sP_%sL.txt" % (P_in,L_in),(N_l,Ar, Vr))
#print(Ar, Vr)


exit()

Exc3/paired_correlation_euste.py
__author__ = "JL Euste"

import numpy as np
from scipy.spatial.distance import pdist
import matplotlib.pyplot as plt


# =============================================================================
# Extract data from HISTORY FILE
# =============================================================================

start_init,end_init = 7,41 #start and end line numbers for initial frame
start_fin = 7967 # starting line number for the final frame
spacing = 6 #line spacing of coordinates of O
end_all = 8001 #last line in the file

coords_o_init,coords_o = [],[]
 
file = open('HISTORY') #open file
for i in range(end_all):
    line = file.readline()
    
    #extract coordinates of oxygen from the initial frame
    if i in np.arange(start_init,end_init,spacing): 
        coords_o_init.append(line)   
        
    #extract coordinates of oxygen from the final frame
    if i in np.arange(start_fin,start_fin+(end_init-start_init),spacing):        
        coords_o.append(line)

def coords_to_array(coords_list):
    ''' Change list of coordinates to array. '''
    n = len(coords_list)
    x = [float(coords_list[i][1:12]) for i in range(n)]
    y = [float(coords_list[i][14:24]) for i in range(n)]
    z = [float(coords_list[i][26:37]) for i in range(n)]  
    return np.array([x,y,z]).T
                
                
coords_o_init = coords_to_array(coords_o_init)
coords_o = coords_to_array(coords_o)


# =============================================================================
# Visualize
# =============================================================================
fig = plt.figure()
ax = fig.add_subplot(projection='3d')
ax.plot(coords_o_init[:,0],coords_o_init[:,1],coords_o_init[:,2],
        'o',label='initial')
ax.plot(coords_o[:,0],coords_o[:,1],coords_o[:,2],
        'o',label='final')
ax.legend()
plt.savefig('oxygen_init_vs_fin',bbox_inches='tight',dpi=300)
plt.show()
plt.close()

# =============================================================================
# Calculate pairwise distances using minimum image convention
# =============================================================================

def min_image(u,v,L=25.20):
    uv = u-v
    uv = uv - np.round(uv/L)*L
    return np.sqrt(np.sum(uv**2))

distances_o_init = pdist(coords_o_init,metric=min_image)
distances_o = pdist(coords_o,metric=min_image)

# =============================================================================
# Plotting
# =============================================================================

plt.hist(distances_o_init,density=True,label='initial')
plt.hist(distances_o,density=True,label='final')
plt.legend()
plt.xlabel('pairwise distance')
plt.ylabel('probability')
plt.savefig('radial_dist_oxygen_init_vs_fin',bbox_inches='tight',dpi=300)
plt.show()


